go-controller/pkg/ovn/multipolicy_test.go

```diff
@@ -0,0 +1,487 @@
+	for i, policytype := range policy.Spec.PolicyTypes {
+		mpolicy.Spec.PolicyTypes[i] = mnpapi.MultiPolicyType(policytype)
+	}
+	return &mpolicy
+}
+
+func (p testPod) addNetwork(netName, nadName, nodeSubnet, nodeMgtIP, nodeGWIP, podIP, podMAC string) {
+	podInfo, ok := p.secondaryPodInfos[netName]
+	if !ok {
+		podInfo = &secondaryPodInfo{
+			nodeSubnet:  nodeSubnet,
+			nodeMgtIP:   nodeMgtIP,
+			nodeGWIP:    nodeGWIP,
+			allportInfo: map[string]portInfo{},
+		}
+		p.secondaryPodInfos[netName] = podInfo
+	}
+	portName := util.GetSecondaryNetworkLogicalPortName(p.namespace, p.podName, nadName)
+	podInfo.allportInfo[nadName] = portInfo{
+		portUUID: portName + "-UUID",
+		podIP:    podIP,
+		podMAC:   podMAC,
+		portName: portName,
+	}
+}
+
+func (p testPod) getNetworkPortInfo(netName, nadName string) *portInfo {
+	podInfo, ok := p.secondaryPodInfos[netName]
+	if !ok {
+		return nil
+	}
+	info, ok := podInfo.allportInfo[nadName]
+	if !ok {
+		return nil
+	}
+
+	return &info
+}
+
+func addPodNetwork(pod *v1.Pod, secondaryPodInfos map[string]*secondaryPodInfo) {
+	nadNames := []string{}
+	for _, podInfo := range secondaryPodInfos {
+		for nadName := range podInfo.allportInfo {
+			nadNames = append(nadNames, nadName)
+		}
+	}
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+	pod.Annotations[nettypes.NetworkAttachmentAnnot] = strings.Join(nadNames, ",")
+}
+
+func (p testPod) populateSecondaryNetworkLogicalSwitchCache(fakeOvn *FakeOVN, ocInfo secondaryControllerInfo) {
+	var err error
+	switch ocInfo.TopologyType() {
+	case ovntypes.Layer3Topology:
+		podInfo := p.secondaryPodInfos[ocInfo.GetNetworkName()]
+		uuid := getLogicalSwitchUUID(fakeOvn.controller.nbClient, ocInfo.GetNetworkScopedName(p.nodeName))
+		err = ocInfo.BaseSecondaryNetworkController.lsManager.AddSwitch(ocInfo.GetNetworkScopedName(p.nodeName), uuid, []*net.IPNet{ovntest.MustParseIPNet(podInfo.nodeSubnet)})
+	case ovntypes.Layer2Topology:
+		uuid := getLogicalSwitchUUID(fakeOvn.controller.nbClient, ocInfo.GetNetworkScopedName(ovntypes.OVNLayer2Switch))
+		subnet := ocInfo.NetConfInfo.Subnets()[0]
+		err = ocInfo.BaseSecondaryNetworkController.lsManager.AddSwitch(ocInfo.GetNetworkScopedName(ovntypes.OVNLayer2Switch), uuid, []*net.IPNet{ovntest.MustParseIPNet(subnet)})
+	case ovntypes.LocalnetTopology:
+		uuid := getLogicalSwitchUUID(fakeOvn.controller.nbClient, ocInfo.GetNetworkScopedName(ovntypes.OVNLayer2Switch))
+		subnet := ocInfo.NetConfInfo.Subnets()[0]
+		err = ocInfo.BaseSecondaryNetworkController.lsManager.AddSwitch(ocInfo.GetNetworkScopedName(ovntypes.OVNLocalnetSwitch), uuid, []*net.IPNet{ovntest.MustParseIPNet(subnet)})
+	}
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+}
+
+func getExpectedDataPodsAndSwitchesForSecondaryNetwork(fakeOvn *FakeOVN, pods []testPod) []libovsdb.TestData {
+	data := []libovsdb.TestData{}
+	for _, ocInfo := range fakeOvn.secondaryControllers {
```

Comment:
I am getting several warnings around this file that we are copying BaseNetworkController namespacesMutex which is not allowed and might not work as expected. I guess you should use pointers to the controllers.